### PR TITLE
Add Sentry to the Calypso client

### PR DESCRIPTION
**Pull request by noahtallen**
### Changes proposed in this Pull Request

Enable sentry error logging in the Calypso client. I think it should also be relatively easy (compared to the wp-admin project, anyways) to enable source map uploads. We can very easily set the release to something like `calypso-$git_hash`, using the existing `window.COMMIT_SHA` property. (Though, a quick test shows we may need to fix that property in production, as it currently returns "(unknown)".) And we could even use the webpack sentry integration to upload source maps from TeamCity.

I copied @sirreal's async-loader for Sentry from Tumblr, which was a huge help! Thanks for the tip. I've configured things a bit differently for Calypso, but overall it's very similar.

To get parity with the existing solution, we should support the following. There is existing art for each of these in the same function I put the sentry handler.
 
- [x] Setting the blog ID
- [x] Setting the Calypso section
- [x] Setting the path in Calypso (e.g. "/home/:siteId" after replacing the actual site URL)

The latter three need to update as the user navigates around Calypso, which means listening to redux state changes. 

You can see the blog ID, user ID, and Calypso section are saved as tags here:
Screen Shot 2022-04-08 at 4 50 19 PM

And here you can see the site slug normalization at work in navigation events:
Screen Shot 2022-04-08 at 4 50 02 PM

### TODO:
- [ ] Figure out how to always enable sentry when proxied for easier testing. (It would also be helpful to specifically tag errors if they're proxied so that we understand if an error is only happening internally.)
- [ ] Fix the COMMIT_SHA property. I think it isn't getting passed into the Docker image somehow.
- [ ] Explore release management with Webpack and TeamCity.
- [ ] Explore if we need to do additional work to enable Sentry on signup pages.

### Testing instructions
- TBD

### CODE REVIEW

File: client/boot/common.js [hunk 1]
```diff
@@ -241,6 +242,18 @@ function setupErrorLogger( reduxStore ) {
...
+	Sentry.init( {
+		dsn: 'https://61275d63a504465ab315245f1a379dab@o248881.ingest.sentry.io/6313676',
+		environment: config( 'env_id' ),
+		initialScope: {
+			user: { id: getCurrentUserId( reduxStore.getState() ) },
+		},
+	} );
+	if ( window._jsErr ) {
+		window._jsErr.forEach( ( error ) => Sentry.captureException( error ) );
+		Sentry.flush().then( () => delete window._jsErr );
```
Comment (noahtallen) [inline client/boot/common.js:254]: Copied this from the existing ETK solution, but I'm not sure it makes sense to delete the error array -- unless we delete the event listener as well, the array would just be recreated.

File: apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js [hunk 2]
```diff
@@ -24,7 +24,10 @@ function activateSentry() {
 
 	// We still need to report the head errors, if any.
 	headErrors.forEach( ( error ) => Sentry.captureException( error ) );
-	Sentry.flush().then( () => delete window._jsErr );
+	Sentry.flush().then( () => {
+		delete window._jsErr;
+		window.removeEventListener( window._headJsErrorHandler );
```
Comment (noahtallen) [inline apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js]: Since the event listener also creates a new jsErr array when it doesn't exist, future onerror events could still be saved, causing the array to become very large

Comment (fullofcaffeine) [reply apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js]: Ah, the actual listener can re-initialize `window._jsErr`... good catch!

**Comment by noahtallen** [comment]
An interesting observation is that we can see the e2e users triggering alerts in sentry! Not sure if we want that behavior. 

We do set the environment -- calypso.live is "wpcalypso" -- so we could just filter by production. But we might end up with lots of extra errors we'll have to resolve for WIP code 🤔

**Comment by fullofcaffeine** [comment]
> An interesting observation is that we can see the e2e users triggering alerts in sentry! Not sure if we want that behavior.

Hmm, isn't that the case now? I mean, E2E users are regular production users/sites (though there are some exceptions based on the user/site name in some places, AFAIU). I didn't put much thought into it when we activated the error reporting the first time or when we added Sentry as I thought that since those were regular production users, any Sentry errors would be legit ones that we should deal with anyways. Or maybe you're talking about a slightly different thing?

By the way, are you planning to enable Sentry for all users or go with a segmented/gradual rollout?

**Comment by noahtallen** [comment]
Yeah, I think it's the case now -- though dashboards might only show the production errors by filtering based on environment.

I think the tradeoff is that events captured when testing a pull request might just be indicating that the PR is buggy. For example if it cashes when you push the PR for the first time, those will be logged to Sentry -- but they aren't really actionable bugs that we need to address unless they go to production.

> By the way, are you planning to enable Sentry for all users or go with a segmented/gradual rollout?

I'd like to segment it before merging, but I haven't looked into how to do that yet

File: client/boot/common.js [hunk 3]
```diff
@@ -264,10 +296,22 @@ function setupErrorLogger( reduxStore ) {
...
-			context.canonicalPath.replace( getSiteFragment( context.canonicalPath ), ':siteId' )
+		const path = context.canonicalPath.replace(
+			getSiteFragment( context.canonicalPath ),
+			':siteId'
 		);
+		Sentry.addBreadcrumb( {
+			type: 'navigation',
+			data: {
+				from: prevPath ?? path,
+				to: path,
```
Comment (noahtallen) [inline client/boot/common.js:297]: While the default navigation handlers do include the path, it also includes the site ID. Following how it worked previously, it makes it easier to work with when we normalize the URL -- so instead of `/home/example.wordpress.com`, it should show `/home/:siteId`, just like logstash.

**Comment by noahtallen** [comment]
23K increase in the main entry point is also a fair size increase. I wonder if we can load sentry only when the error reporting segment is active -- maybe like for 20% of calypso users?

File: apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js (hunk 2 above)
Comment (noahtallen) [reply apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js]: Actually, my mistake! The script already deletes the event listener, just at the very end of the file.

File: client/boot/common.js [hunk 4]
```diff
@@ -237,10 +238,27 @@ const configureReduxStore = ( currentUser, reduxStore ) => {
...
+	// Note that Sentry can disable itself and do some cleanup if needed, so we
+	// run it before the catch-js-errors check. (Otherwise, cleanup would never
+	// never happen.)
+	initSentry( { beforeSend } );
+
 	if ( ! config.isEnabled( 'catch-js-errors' ) ) {
 		return;
 	}
 
+	// At this point, the normal error logger is still set up so that
```
Comment (fullofcaffeine) [inline client/boot/common.js]: Looks like this sentence needs to be finished after "so that..."?

File: client/boot/common.js [hunk 5]
```diff
@@ -264,10 +282,23 @@ function setupErrorLogger( reduxStore ) {
...
+			getSiteFragment( context.canonicalPath ),
+			':siteId'
 		);
+		// Also save the context to Sentry for easier debugging.
+		addBreadcrumb( {
+			category: 'navigation',
+			data: {
+				from: prevPath ?? path,
+				to: path,
+				should_capture: true, // Hint that this is our own breadcrumb, not the default navigation one.
```
Comment (fullofcaffeine) [inline client/boot/common.js:298]: What is the practical effect of that in Sentry?

File: client/boot/common.js [hunk 6]
```diff
@@ -264,10 +282,23 @@ function setupErrorLogger( reduxStore ) {
 		errorLogger.saveExtraData( { lastTracksEvent } )
 	);
 
+	let prevPath;
```
Comment (fullofcaffeine) [inline client/boot/common.js:286]: Could you elaborate on the logic that relies on this variable?

File: client/boot/common.js (hunk 4 above)
Comment (noahtallen) [reply client/boot/common.js]: Good catch!

File: client/boot/common.js (hunk 5 above)
Comment (noahtallen) [reply client/boot/common.js:298]: Right, so this is actually deleted before getting sent through the API. The purpose is to basically indicate that this event is our own custom navigation event. That way, we discard any navigation breadcrumbs tracked by sentry, and only keep breadcrumbs that have this flag on them. This is accomplished with a small amount of logic in the "beforeBreadcrumb" handler.

File: client/lib/sentry/index.ts [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+// purpose is to stub basic Sentry methods so that consumers in Calypso can use
+// them. Then we can async-load Sentry for only a certain percent of requests.
+// Otherwise, we'd add a fair amount to the bundle size when we don't really need
+// it for every single request.
+import config from '@automattic/calypso-config';
+import type * as SentryApi from '@sentry/react';
+
+// Static sentry configuration. We add any dynamic values when initializing Sentry.
+const SENTRY_CONFIG: SentryApi.BrowserOptions = {
+	dsn: 'https://61275d63a504465ab315245f1a379dab@o248881.ingest.sentry.io/6313676',
```
Comment (fullofcaffeine) [inline client/lib/sentry/index.ts:11]: Do you think we should create a new Sentry project for Calypso?

Comment (fullofcaffeine) [reply client/lib/sentry/index.ts:11]: Ah, disregard this. I see it's already there :)

File: client/boot/common.js (hunk 6 above)
Comment (noahtallen) [reply client/boot/common.js:286]: Yes, so sentry navigation breadcrumbs need to have a properties for both the page you navigate _from_ and the page you navigate _to_. Keeping track of the `prevPath` means we can maintain the "from" data, and then the "to" data is just the current page in the page navigation handler below.

We just update this variable after sending the breadcrumb to sentry with the new "from" page, so that it's available for the next event.

File: client/lib/sentry/index.ts (hunk 7 above)
Comment (noahtallen) [reply client/lib/sentry/index.ts:11]: This is the new project for Calypso (https://sentry.io/organizations/a8c/projects/calypso/?project=6313676)

File: client/lib/sentry/index.ts [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+			return;
+		}
+		state = { state: 'loading' };
+
+		// Enable Sentry only for 10% of requests or always in calypso.live for testing.
+		// Always disable if catch-js-errors is not available in the environment.
+		if (
+			! (
+				config.isEnabled( 'catch-js-errors' ) &&
+				( config( 'env_id' ) === 'wpcalypso' || Math.floor( Math.random() * 10 ) === 1 )
```
Comment (fullofcaffeine) [inline client/lib/sentry/index.ts:134]: Curious - where is the `env_id` set?

Just a nit and no need to act on that but, a better value for `calypso.live` would probably be `calypso.live`? :)

Comment (noahtallen) [reply client/lib/sentry/index.ts:134]: Everything from `config` comes from all the normal config files here: https://github.com/Automattic/wp-calypso/tree/trunk/config. Basically different data for the different environments :) I agree that the name is confusing, but I think it's been this way for long enough that we might break things if we change it :(

File: client/boot/common.js (hunk 5 above)
Comment (fullofcaffeine) [reply client/boot/common.js:298]: Ah, makes sense 👍🏻

File: client/lib/sentry/index.ts (hunk 8 above)
Comment (fullofcaffeine) [reply client/lib/sentry/index.ts:134]: Thanks for pointing that out :)

File: client/boot/common.js (hunk 6 above)
Comment (fullofcaffeine) [reply client/boot/common.js:286]: Ah, I see. I was initially confused because I thought this script ran in the SSR context, which would mean this variable would be shared across requests. Now I see it's the *client*-side route handler, and now it makes sense!

File: client/boot/common.js [hunk 9]
```diff
@@ -237,10 +238,28 @@ const configureReduxStore = ( currentUser, reduxStore ) => {
...
+		}
+		event.tags.blog_id = getSelectedSiteId( state );
+		event.tags.calypso_section = getSectionName( state );
+		return event;
+	};
+
+	// Note that Sentry can disable itself and do some cleanup if needed, so we
+	// run it before the catch-js-errors check. (Otherwise, cleanup would never
+	// never happen.)
+	initSentry( { beforeSend } );
```
Comment (fullofcaffeine) [inline client/boot/common.js:255]: Should we move this call after the condition that checks if the `catch-js-errors` config is `true`? It doesn't seem to make sense to init Sentry here and `return` just afterwards, while there's still code that should run that Sentry depends on (the one that's part of the `page('*')` route, so I assume we should *not* init Sentry if `catch-js-errors` is `false`.

**Comment by fullofcaffeine** [comment]
I'd like to spend a bit more time testing it tomorrow, before giving the final 👍🏻, but overall this is looking good! 🎉 Thanks a lot for working on this! 🙇🏻

**Review by fullofcaffeine — APPROVED** [review]
~There's a [comment](https://github.com/Automattic/wp-calypso/pull/62643/files#r854689248) I'd like to follow up on, but overall this~ looks good and tests well! 🎉

File: client/boot/common.js (hunk 9 above)
Comment (noahtallen) [reply client/boot/common.js:255]: I tried to explain in the comment, but it could definitely be clearer! More specifically, the way the async-loaded sentry package works, anything in Calypso should be able to import and call our stubbed Sentry functions. This way, we can utilize sentry functions like "add breadcrumb" anywhere in the Calypso application without wrapping it in a conditional -- because as we know, Sentry might not always be loaded.

Internally, those functions will queue up calls to sentry as long as sentry is not loaded. As a result, if we know we're _never_ going to load sentry, that queue could (theoretically) get pretty big because the queue is never processed. So if we know we aren't going to use sentry at all, those stubbed functions should just return and not do anything -- and the queue of calls should be cleared once we know sentry won't be loaded.

When `initSentry` is called, it checks all of the factors -- such as the RNG, `catch-js-errors`, etc. If it determines we shouldn't load sentry, it clears the queues and exits. It also sets its status to "disabled", which causes the stubbed functions to just return without doing anything. So at that point, the stubbed functions can be called anywhere in Calypso, but Sentry isn't loaded and they don't do anything, minimizing the bundle size impact.

In other words, it's designed to called _always_, and it conditionally disables itself internally.

I could also rewrite this to be more explicit if the approach isn't very clear 🤔

Comment (fullofcaffeine) [reply client/boot/common.js:255]: Ah, I see, thanks for the comprehensive explanation! 👍🏻
